File: web/src/lib/tail.ts
/** Overlap-dedup follow for snapshot log endpoints (PU4-6 — U3 tail.py).
 *
 * The daemon exposes logs as "last N lines" snapshots; `advance` computes the
 * newly appended suffix between two snapshots by matching the largest overlap
 * between the previous snapshot's tail and the new snapshot's head. A vanished
 * overlap (rotation/window overrun) is reported as a gap, never silently
 * skipped.
 */

export interface TailStep {
  newLines: string[]
  gap: boolean
}

export function advance(prev: readonly string[], fetched: readonly string[]): TailStep {
  if (prev.length === 0) return { newLines: [...fetched], gap: false }
  if (fetched.length === 0) return { newLines: [], gap: false }
  const maxK = Math.min(prev.length, fetched.length)
  for (let k = maxK; k > 0; k--) {
    let match = true
    for (let i = 0; i < k; i++) {
      if (prev[prev.length - k + i] !== fetched[i]) {
        match = false
        break
      }
    }
    if (match) return { newLines: fetched.slice(k), gap: false }
  }
  return { newLines: [...fetched], gap: true }
}

File: caduceus/cli/tail.py
"""Overlap-dedup follow for snapshot log endpoints (pattern P8, PU3-6).

The daemon exposes logs as "last N lines" snapshots; ``advance`` computes the
newly appended suffix between two snapshots by matching the largest overlap
between the previous snapshot's tail and the new snapshot's head.

Known limitation (documented, business-rules CLI-E2 no-silent-failure): with
fully repetitive identical lines the overlap is ambiguous and the largest-match
rule may conservatively dedup; gateway logs are timestamped so this does not
occur in practice. A vanished overlap (rotation/gap) is reported, never
silently skipped.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class TailStep:
    new_lines: list[str]
    gap: bool  # True when no overlap was found (rotation or window overrun)


def advance(prev: list[str], fetched: list[str]) -> TailStep:
    """Pure: previous snapshot × new snapshot → newly appended lines."""
    if not prev:
        return TailStep(new_lines=list(fetched), gap=False)
    if not fetched:
        return TailStep(new_lines=[], gap=False)
    max_k = min(len(prev), len(fetched))
    for k in range(max_k, 0, -1):
        if prev[len(prev) - k:] == fetched[:k]:
            return TailStep(new_lines=list(fetched[k:]), gap=False)
    return TailStep(new_lines=list(fetched), gap=True)
